PR and Issue templates

ibrahimlawal opened the pull request:
Created Pull-Request and Issue templates that may be used in projects. Added them in a `.github` folder.  Based on recent update from github [here](https://github.com/blog/2111-issue-and-pull-request-templates)

shadowhand commented:
I would be more comfortable with a PR template like this:

```md
Why was this change made?

Fixes #?
Refs #?
```

philsturgeon commented on .github/ISSUE_TEMPLATE.md:
```diff
@@ -0,0 +1,21 @@
+2. Step 2
+3. ...
+
+## Expected Results:
```
+1 on these. 

Basically, if we can shrink advice here (http://www.chiark.greenend.org.uk/~sgtatham/bugs.html) down and get it into a template, that'd be great. This article is perfect.

ibrahimlawal replied:
Will give some thought to that. Read through and did find it awesome.

ibrahimlawal commented:
@shadowhand I believe the committer should also be able to explain changes made. Many times just seeing what was typed doesn't give one a complete idea of its implications. You are right about putting the reason first. A simple patch correcting version number etc need not reference anything. Thought to put `Fix #?` first so that the committers knows to reference issues off the bat.

shadowhand commented:
@ibrahimlawal the committer needs to say _why_ a change was made, not _what_ changed. The _what_ is the diff, a value that is already known.

ibrahimlawal commented:
I love using samples to explain my point.

Here's a sample pull request titled: Added ability to load dotenv files.
```
### Why was this change made?
Using a `.env` file makes it easy ...

### Why was each change made?
- .env.sample was added to provide a sample .env that could be moved by user
- vlucas/phpdotenv require was added in composer.json - This is the plugin that loads .env files into the php `$_ENV` variable
- added .env to .gitignore so it will never get added to the repo
- Edited README to include steps for setting up .env on system

Fixes #xxx
Ref #xxx
```

In this trivial example, were this the first time the repo owner had seen the .env integration, he'd probably question adding `.env` to .gitignore were there no explanation as to why. Having it specified that a _why_ be specified for each change to each file would make the submitter be sure they actually need that change in this commit. is what I think.

summary: I think instead of asking only:
`Why was this change made?`
we should also make the committer explain:
`Why was each change made?`

shadowhand commented:
Again, I don't agree. This is an unnecessary burden on the PR creator. If the reviewer cannot infer what the creator intended simply by looking at the description and the diff, then the PR is not of high quality.

Your "why each change was made" is mostly a *what* list. It doesn't help me understand the code. If any of these things are useful to the next person editing the code (for example, the note about `.env` being excluded) then it should be contained in the code.

I will also mention http://chris.beams.io/posts/git-commit/

ibrahimlawal commented:
Would you rather a note letting the PR creator know that their commit will be rejected if their commit messages do not meet certain standards. And then point them to the article?

shadowhand commented:
I would generally agree with that approach.

philsturgeon commented:
We need to be sure about what level we're looking to have the most clarity possible. If we're going to be asking that people put a lot of effort into their commit messages, then PR content is not so important. 

If people are making PRs with multiple commits, which are each documented excellently, then asking them to write up a fancy PR template is potentially wasting their time. 

If people are making PRs with single commits, then they don't need to be quite so wordy on the commits, because GitHub will link the PR in place, which will contain further documentation. 

Which would we prefer?

I'm in the camp of "Keep your commit messages simple, but give PRs a little more context", and then "One commit per PR". That's what I've done at the last two jobs and its been rather nice. Rebasing all those nasty commits into one when its ready to merge.

That way your commits can be:

> Added some X to a Y
>  
> Without X, Y was all shitty and Z was breaking down why sdfjhlsdhfjlsdkfnadf

Then you can copy that Why into the PR as party of the larger template you're writing. 

Do we agree on that bit? 

Otherwise we're gonna chase each other in circles with words for a bit. Gotta take some middle ground. :)

shadowhand commented:
I'm firmly in the camp of "better commit" because commits are git history, distributed with every single copy of the repository. The PR is relevant **only** to Github and is not accessible directly in the repo.

I too prefer single commit PRs, but only when working with smaller fixes. A commit should not be constrained based on *size* but on being a *unit of work*.

philsturgeon commented:
@shadowhand I get that, and I'm not super happy to rely on a vendor, but we can only provide default PR/issue templates, not commit templates.

We can suggest that people make a What commit message (title) and a few sentences for Why on the commit body. Maybe that can be in the template, that the problem/resolution should be included, but the rest of this stuff can only be in the PR.

I don't want to try and make people put aaaaalll of this stuff back into their commits after the fact. 

If that's not what you're suggesting then :speak_no_evil:

shadowhand commented:
@philsturgeon we can, in fact, provide recommendations for commit templates https://github.com/sjparkinson/static-review#for-commit-messages

philsturgeon commented:
Done in #89! Thanks for the conversation on this one folks, but uncertainty stalled us out, and these new ones are cool.
